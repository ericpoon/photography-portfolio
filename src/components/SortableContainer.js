import React from 'react';
import { DragSource, DropTarget } from 'react-dnd';
import flow from 'lodash/flow';

const itemType = 'SORTABLE';

class SortableContainer extends React.Component {
  render() {
    const { connectDragSource, connectDropTarget, children } = this.props;
    return connectDragSource(connectDropTarget(
      <div
        ref={(node) => {
          this.node = node;
        }}
      >
        {children}
      </div>
    ));
  }
}

const cardSource = {
  beginDrag(props) {
    let { index } = props;
    return {
      get index() {
        return index;
      },
      setIndex(newIndex) {
        index = newIndex;
      },
    };
  },
};

const cardTarget = {
  hover(props, monitor, component) {
    const { decoratedComponentInstance } = component;
    const dragIndex = monitor.getItem().index;
    const hoverIndex = props.index;

    if (dragIndex === hoverIndex) return;

    // Determine rectangle on screen, avoid using findDOMNode
    const hoverBoundingRect = decoratedComponentInstance.node.getBoundingClientRect();

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    if (clientOffset.x > hoverBoundingRect.left
      && clientOffset.x < hoverBoundingRect.right
      && clientOffset.y > hoverBoundingRect.top
      && clientOffset.y < hoverBoundingRect.bottom) {
      props.trigger(dragIndex, hoverIndex); // actually perform the action
      monitor.getItem().setIndex(hoverIndex); // update the item being dragged
    }
  },
};

const sourceCollect = (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  isDragging: monitor.isDragging(),
});
const targetCollect = (connect, monitor) => ({
  connectDropTarget: connect.dropTarget(),
});

export default flow(
  DragSource(itemType, cardSource, sourceCollect),
  DropTarget(itemType, cardTarget, targetCollect)
)(SortableContainer);
