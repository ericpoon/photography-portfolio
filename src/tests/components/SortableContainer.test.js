import React from 'react';
import { shallow } from 'enzyme';
import SortableContainer from '../../components/SortableContainer';

describe('Testing the original component in isolation', () => {
  let wrapper,
    connectDragSource,
    connectDropTarget;
  beforeEach(() => {
    const { DecoratedComponent } = SortableContainer;
    connectDragSource = jest.fn(i => i);
    connectDropTarget = jest.fn(i => i);
    wrapper = shallow(
      <DecoratedComponent
        connectDragSource={connectDragSource}
        connectDropTarget={connectDropTarget}
      />
    );
  });

  it('should render container correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should be both a drag source and a drop target', () => {
    expect(connectDragSource).toHaveBeenCalled();
    expect(connectDropTarget).toHaveBeenCalled();
  });
});

describe('Testing interaction', () => {

});
