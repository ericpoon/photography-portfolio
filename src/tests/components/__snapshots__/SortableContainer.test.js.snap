// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the original component in isolation should render container correctly 1`] = `<div />`;
